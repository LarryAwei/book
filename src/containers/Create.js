import React from 'react'
import CategorySelect from "../components/CategorySelect";
import PriceForm from "../components/PriceForm";
import {Tabs, Tab} from "../components/Tabs";
import {TYPE_OUTCOME, TYPE_INCOME} from "../utility";
import {AppContext} from "../App";
import {testCategories} from "../testData";
import {withContext} from "../WithContext";

const tabsText = [TYPE_OUTCOME, TYPE_INCOME]
export class Create extends React.Component {
    // constructor(props) {
    //     super(props)
    //     this.state = {
    //         selectedTab: TYPE_OUTCOME,
    //         selectedCategory: null,
    //         validationPassed: true,
    //     }
    // }
    // componentDidMount() {
    //     const { id } = this.props.match.params
    //     this.props.actions.getEditData(id).then(data => {
    //         const { editItem, categories } = data
    //         this.setState({
    //             selectedTab: (id && editItem) ? categories[editItem.cid].type : TYPE_OUTCOME,
    //             selectedCategory: (id && editItem) ? categories[editItem.cid] : null,
    //         })
    //     })
    // }
    tabChange = (index) => {
        this.setState({
            selectedTab: tabsText[index]
        })
    }
    selectCategory = (category) => {
        this.setState({
            selectedCategory: category
        })
    }
    cancelSubmit = () => {
        this.props.history.push('/')
    }
    submitForm = (data, isEditMode) => {
        if (!this.state.selectedCategory) {
            this.setState({
                validationPassed: false
            })
            return
        }
        if(!isEditMode) {
            // create
            this.props.actions.createItem(data, this.state.selectedCategory.id).then(this.navigateToHome)
        } else {
            // update
            this.props.actions.updateItem(data, this.state.selectedCategory.id).then(this.navigateToHome)
        }
        this.props.history.push('/')

    }
    navigateToHome = () => {
        this.props.history.push('/')
    }
    render() {
        const {data} = this.props
        console.log(data)
        const filterCategories = testCategories.filter(category=>category.type===TYPE_OUTCOME)
        // const { data } = this.props
        // const { items, categories } = data
        // const { id } = this.props.match.params
        // const editItem = (id && items[id]) ? items[id] : {}
        // const { selectedTab, selectedCategory, validationPassed } = this.state
        // const filterCategories = Object.keys(categories)
        //     .filter(id => categories[id].type === selectedTab)
        //     .map(id => categories[id])
        // const tabIndex = tabsText.findIndex(text => text === selectedTab)
        return (


                        <div className="create-page py-3 px-3 rounded mt-3" style={{background: '#fff'}}>
                            {/*{ data.isLoading &&*/}
                            {/*<Loader />*/}
                            {/*}*/}
                            <Tabs activeIndex={0} onTabChange={this.tabChange}>
                                <Tab>支出</Tab>
                                <Tab>收入</Tab>
                            </Tabs>
                            <CategorySelect categories={filterCategories}
                                            onSelectCategory={this.selectCategory}
                                //selectedCategory={selectedCategory}
                            />
                            <PriceForm
                                onFormSubmit={this.submitForm}
                                onCancelSubmit={this.cancelSubmit}
                                // item={editItem}
                            />
                            {/*{ !validationPassed &&*/}
                            {/*<div className="alert alert-danger mt-5" role="alert">*/}
                            {/*    请选择分类信息*/}
                            {/*</div>*/}
                            {/*}*/}
                        </div>

        )
    }
}

// CreatePage.propTypes = {
//     data: PropTypes.object.isRequired,
//     actions: PropTypes.object.isRequired,
//     history: PropTypes.object,
//     match: PropTypes.object,
// }

//export default withRouter(withContext(CreatePage))
export default withContext(Create)
