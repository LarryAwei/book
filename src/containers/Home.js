import React, {Component} from 'react'
import {LIST_VIEW, CHART_VIEW, TYPE_INCOME, TYPE_OUTCOME, parseToYearAndMonth, padLeft} from "../utility";
import PriceList from "../components/PriceList";
import ViewTab from "../components/ViewTab";
import TotalPrice from "../components/TotalPrice";
import MonthPicker from "../components/MonthPicker";
import CreateBtn from "../components/CreateBtn";
import {Tabs, Tab} from "../components/Tabs";
import {flatternArr} from "../utility";
import Ionicon from "react-ionicons";
import {testCategories, testItems} from "../testData";
import {AppContext} from "../App";
import {withContext} from "../WithContext";
import {withRouter} from 'react-router-dom'

// export const categories = {
//     "1": {
//         "id": "1",
//         "name": "旅行",
//         "type": "outcome",
//         "iconName": "ios-plane"
//     },
//     "2": {
//         "id": "2",
//         "name": "红包",
//         "type": "income",
//         "iconName": "ios-plane"
//     }
// };
// export const items = [
//     {
//         "id": 1,
//         "title": "旅游",
//         "price": 200,
//         "date": "2019-10-20",
//         "cid" : 1
//     },
//     {
//         "id": 2,
//         "title": "充值",
//         "price": 100,
//         "date": "2019-12-20",
//         "cid": 2
//     }
// ];
const newItem = {
    "id": 4,
    "title": "新项目",
    "price": 3000,
    "date": "2019-09-12",
    cid: 1
}
class Home extends Component{
    constructor(props){
        super(props);
        this.state = {
            items: testItems,
            currentDate: parseToYearAndMonth(),
            tabView: LIST_VIEW
        }
    }

    changeView = (view) => {
        this.setState({
            tabView: view
        })
    }
    changeDate = (year, month) => {
        this.setState({
            currentDate: {year, month}
        })

    }
    createItem = () => {
        this.props.history.push('/create')
        // this.setState({
        //     items: [newItem, ...this.state.items]
        // })
    }
    modifyItem = (mitem) => {
        this.props.history.push(`/edit/${mitem.id}`)
        // const modifyItems = this.state.items.map(item=>{
        //     if(item.id===mitem.id)
        //         return {...item, title: 'new'}
        //     else
        //         return item
        // })
        // this.setState({
        //     items: modifyItems
        // })
    }
    deleteItem = (ditem) => {
        this.props.actions.deleteItem(ditem)
        // const filterItems = this.state.items.filter(item=>item.id!==ditem.id)
        // this.setState({
        //     items: filterItems
        // })
    }
    render() {
        const {data} = this.props
        const {items, currentDate, tabView} = this.state
        const itemWithCategory = items.map(item=>{
            item.category=flatternArr(testCategories)[item.cid];
            return item
        }).filter(item=>{
            return item.date.includes(`${currentDate.year}-${padLeft(currentDate.month)}`)
        })

        let totalIncome = 0, totalOutcome = 0
        itemWithCategory.forEach(item => {
            if (testCategories[0].type === TYPE_OUTCOME) {
                totalOutcome += item.price
            } else {
                totalIncome += item.price
            }
        });
        return(

                        <React.Fragment>
                            <header className="App-header">
                                <p>
                                    Edit <code>src/App.js</code> and save to reload.
                                </p>
                                <a
                                    className="App-link"
                                    href="https://reactjs.org"
                                    target="_blank"
                                    rel="noopener noreferrer"
                                >
                                    Learn React
                                </a>
                                <div className="row">
                                    <div className="col">
                                        <MonthPicker
                                            year={currentDate.year}
                                            month={currentDate.month}
                                            onChange={this.changeDate}
                                        />
                                    </div>
                                    <div className="col">
                                        <TotalPrice
                                            income={totalIncome}
                                            outcome={totalOutcome}
                                        />
                                    </div>
                                </div>

                            </header>
                            <div className={"content-area py-3 px-3"}>
                                <Tabs activeIndex={0} onTabChange={this.changeView}>
                                    <Tab>
                                        <Ionicon
                                            className="rounded-circle mr-2"
                                            fontSize="25px"
                                            color={'#007BFF'}
                                            icon='ios-paper'
                                        />
                                        列表模式
                                    </Tab>
                                    <Tab>
                                        <Ionicon
                                            className="rounded-circle mr-2"
                                            fontSize="25px"
                                            color={'#007BFF'}
                                            icon='ios-pie'
                                        />
                                        图表模式
                                    </Tab>
                                </Tabs>
                                {/*<ViewTab activeTab={tabView} onTabChange={this.changeView} />*/}
                                <CreateBtn  onClick={this.createItem}/>
                                {tabView===LIST_VIEW &&
                                <PriceList items={itemWithCategory} onModifyItem={this.modifyItem} onDelItem={this.deleteItem}/>
                                }
                                {
                                    tabView===CHART_VIEW &&
                                    <h1>图标</h1>
                                }

                            </div>
                        </React.Fragment>

        )
    }
}
export default withRouter(withContext(Home))
