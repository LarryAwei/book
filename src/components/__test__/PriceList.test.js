import React from 'react'
import {shallow} from 'enzyme'
import Enzyme from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import toJson from "enzyme-to-json";
import PriceList from "../PriceList";
import {items, categories} from '../../containers/Home'
Enzyme.configure({ adapter: new Adapter() });

const itemWithCategory = items.map(item=>{
    item.category = categories[item.cid]
    return item
});
const props = {
    items: itemWithCategory,
    onModifyItem: jest.fn(),
    onDelItem: jest.fn()
};
let wrapper;
describe('test priceList', ()=>{
    beforeEach(()=>{
        wrapper = shallow(<PriceList {...props}/>)
    });
    it('should render a component matching the snapshot', () => {
        const tree = toJson(wrapper);
        expect(tree).toMatchSnapshot();

    });
    it('should length', () => {
        expect(wrapper.find('.list-group-item').length).toEqual(itemWithCategory.length)
    });
    it('should trigger function callback',  () => {
        const firstItem = wrapper.find('.list-group-item').first()
        firstItem.find('a').first().simulate('click')
        expect(props.onModifyItem).toHaveBeenCalledWith(itemWithCategory[0])
    });
});
