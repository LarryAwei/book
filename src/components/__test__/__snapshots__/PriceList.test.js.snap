// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test priceList should render a component matching the snapshot 1`] = `
<ul
  className="list-group list-group-flush"
>
  <li
    className="list-group-item d-flex justify-content-between aligh-items-center"
    key="1"
  >
    <span
      className="col-1 badge badge-primary"
    >
      <Ionicon
        beat={false}
        className="rounded-circle"
        color="#000000"
        fontSize="30px"
        icon="ios-plane"
        rotate={false}
        shake={false}
        style={
          Object {
            "backgroundColor": "#007bff",
            "color": "#fff",
            "padding": "5px",
          }
        }
      />
    </span>
    <span
      className="col-5"
    >
      旅游
    </span>
    <span
      className="col-2 font-weight-bold"
    >
      -
       $
      200
    </span>
    <span
      className="col-2"
    >
      2019-10-20
    </span>
    <a
      className="col-1"
      onClick={[Function]}
    >
      <Ionicon
        beat={false}
        className="rounded-circle"
        color="#000000"
        fontSize="30px"
        icon="ios-create-outline"
        rotate={false}
        shake={false}
        style={
          Object {
            "backgroundColor": "green",
            "color": "#fff",
            "padding": "5px",
          }
        }
      />
    </a>
    <a
      className="col-1"
      onClick={[Function]}
    >
      <Ionicon
        beat={false}
        className="rounded-circle"
        color="#000000"
        fontSize="30px"
        icon="ios-plane"
        rotate={false}
        shake={false}
        style={
          Object {
            "backgroundColor": "red",
            "color": "#fff",
            "padding": "5px",
          }
        }
      />
    </a>
  </li>
  <li
    className="list-group-item d-flex justify-content-between aligh-items-center"
    key="2"
  >
    <span
      className="col-1 badge badge-primary"
    >
      <Ionicon
        beat={false}
        className="rounded-circle"
        color="#000000"
        fontSize="30px"
        icon="ios-plane"
        rotate={false}
        shake={false}
        style={
          Object {
            "backgroundColor": "#007bff",
            "color": "#fff",
            "padding": "5px",
          }
        }
      />
    </span>
    <span
      className="col-5"
    >
      充值
    </span>
    <span
      className="col-2 font-weight-bold"
    >
      +
       $
      100
    </span>
    <span
      className="col-2"
    >
      2019-12-20
    </span>
    <a
      className="col-1"
      onClick={[Function]}
    >
      <Ionicon
        beat={false}
        className="rounded-circle"
        color="#000000"
        fontSize="30px"
        icon="ios-create-outline"
        rotate={false}
        shake={false}
        style={
          Object {
            "backgroundColor": "green",
            "color": "#fff",
            "padding": "5px",
          }
        }
      />
    </a>
    <a
      className="col-1"
      onClick={[Function]}
    >
      <Ionicon
        beat={false}
        className="rounded-circle"
        color="#000000"
        fontSize="30px"
        icon="ios-plane"
        rotate={false}
        shake={false}
        style={
          Object {
            "backgroundColor": "red",
            "color": "#fff",
            "padding": "5px",
          }
        }
      />
    </a>
  </li>
</ul>
`;
